### PR TITLE
[tensorexpr] Move the allocation of intermediate buffers from TEK to CodeGen

### PR DESCRIPTION
**Pull request by huiguoo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66967
* #66452
* #67144
* __->__ #67143
* #66451
* #66450

Differential Revision: [D31881151](https://our.internmc.facebook.com/intern/diff/D31881151)

### CODE REVIEW

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

File: torch/csrc/jit/tensorexpr/codegen.cpp [hunk 1]
```diff
@@ -80,6 +82,50 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
   TORCH_INTERNAL_ASSERT(
       false, "This codegen backend does not implement call_with_numel");
 }
+
+StmtPtr insertMemNodes(std::unordered_set<BufPtr>& interm_bufs, StmtPtr stmt) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/codegen.cpp:86]: Nit: Let's be more precise with names. MemNodes can mean many different things, why do we need to change it from `insertAllocFree`?

**Review by ZolotukhinM — APPROVED** [review]
Looks good!

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31881151).

**Pull request closed without merging**